bob-apps/bin/config_cluster_generator/pattern.rs

```diff
@@ -0,0 +1,250 @@
+use anyhow::{anyhow, Result as AnyResult};
+use bob_common::configs::cluster::Node;
+use bob_common::core_types::{DiskName, DiskPath};
+use itertools::Itertools;
+use regex::{Captures, Regex};
+
+fn parse_address_sample(sample: &str) -> AnyResult<((String, u16), String)> {
+    debug!("Sample to parse: {}", sample);
+    let re = Regex::new(r"^(?<addr>[\w.]+):(?<port>\d+)(?<path>/[\w/]+)$").unwrap();
+    if let Some(captures) = re.captures(sample) {
+        let ip = captures["addr"].to_owned();
+        let port = &captures["port"];
+        let path = captures["path"].to_owned();
+        let port: u16 = port
+            .parse()
+            .map_err(|_| anyhow!("Failed to parse port {}", port))?;
+        Ok(((ip, port), path))
+    } else {
+        Err(anyhow!("Failed to parse the sample {}", sample))
+    }
+}
+
+fn substitute_node_pattern(node_pattern: &str, ip: &str, port: u16, id: usize) -> String {
+    let re = Regex::new(r"\{(\w+)}").unwrap();
+    re.replace_all(node_pattern, |caps: &Captures| match &caps[1] {
+        "ip" => ip.to_string(),
+        "port" => port.to_string(),
+        "id" => id.to_string(),
+        _ => caps[0].to_string(),
+    })
+    .to_string()
+}
+
+fn generate_range_samples(pattern: &str) -> impl Iterator<Item = String> {
+    let re = Regex::new(r"\[(\d+)-(\d+)]").unwrap();
+
+    let ranges = re.captures_iter(pattern).map(|captures| {
+        let start: usize = captures[1].parse().unwrap();
+        let end: usize = captures[2].parse().unwrap();
+        start..=end
+    });
+
+    re.split(pattern)
+        .zip_longest(ranges)
+        .map(|x| {
+            if let itertools::EitherOrBoth::Both(part, range) = x {
+                range.map(|i| part.to_string() + &i.to_string()).collect()
+            } else {
+                vec![x.left().expect("is some because split > range").to_string()]
+            }
+        })
+        .multi_cartesian_product()
+        .map(|x| x.concat())
+        .into_iter()
+}
+
+fn pattern_extend_disks(node: &mut Node, disk_paths: impl Iterator<Item = String>) {
+    let old_disks = node.disks();
+    let new_disks: Vec<DiskPath> = disk_paths
+        .filter(|disk_path| !old_disks.iter().any(|d| d.path() == *disk_path))
```

Comment:
It's better to create set here (`HashSet`) with old disk paths and check if set does not contain path